core/src/layers/capability_check.rs

```diff
@@ -0,0 +1,965 @@
+        let capability = self.info.full_capability();
+        if !capability.list_with_version && args.version() {
+            return Err(CapabilityCheckLayer::new_unsupported_args_error(
+                self.info.as_ref(),
+                Operation::List,
+                "version",
+            ));
+        }
+
+        self.inner.list(path, args).await
+    }
+
+    fn blocking_read(
+        &self,
+        path: &str,
+        args: OpRead,
+    ) -> crate::Result<(RpRead, Self::BlockingReader)> {
+        let capability = self.info.full_capability();
+        if !capability.read_with_version && args.version().is_some() {
+            return Err(CapabilityCheckLayer::new_unsupported_args_error(
+                self.info.as_ref(),
+                Operation::BlockingRead,
+                "version",
+            ));
+        }
+
+        self.inner.blocking_read(path, args)
+    }
+
+    fn blocking_write(
+        &self,
+        path: &str,
+        args: OpWrite,
+    ) -> crate::Result<(RpWrite, Self::BlockingWriter)> {
+        let capability = self.info.full_capability();
+        if !capability.write_with_if_none_match && args.if_none_match().is_some() {
+            return Err(CapabilityCheckLayer::new_unsupported_args_error(
+                self.info.as_ref(),
+                Operation::BlockingWrite,
+                "if_none_match",
+            ));
+        }
+
+        self.inner.blocking_write(path, args)
+    }
+
+    fn blocking_stat(&self, path: &str, args: OpStat) -> crate::Result<RpStat> {
+        let capability = self.info.full_capability();
+        if !capability.stat_with_version && args.version().is_some() {
+            return Err(CapabilityCheckLayer::new_unsupported_args_error(
+                self.info.as_ref(),
+                Operation::BlockingStat,
+                "version",
+            ));
+        }
+
+        self.inner.blocking_stat(path, args)
+    }
+
+    fn blocking_delete(&self, path: &str, args: OpDelete) -> crate::Result<RpDelete> {
+        let capability = self.info.full_capability();
+        if !capability.delete_with_version && args.version().is_some() {
+            return Err(CapabilityCheckLayer::new_unsupported_args_error(
+                self.info.as_ref(),
+                Operation::BlockingDelete,
+                "version",
+            ));
+        }
+
+        self.inner.blocking_delete(path, args)
+    }
+
+    fn blocking_list(
+        &self,
+        path: &str,
+        args: OpList,
+    ) -> crate::Result<(RpList, Self::BlockingLister)> {
+        let capability = self.info.full_capability();
+        if !capability.list_with_version && args.version() {
+            return Err(CapabilityCheckLayer::new_unsupported_args_error(
+                self.info.as_ref(),
+                Operation::BlockingList,
+                "version",
+            ));
+        }
+
+        self.inner.blocking_list(path, args)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Todo: If this solution is confirmed, I'll add tests for the `CorrectnessCapabilityChecker`